File: tests/algorithms/x/Kotlin/maths/addition_without_arithmetic.kt
import java.math.BigInteger

fun _numToStr(v: Number): String {
    val d = v.toDouble()
    val i = d.toLong()
    return if (d == i.toDouble()) i.toString() else d.toString()
}

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

var MAX: Long = 4294967296L
var HALF: Long = 2147483648L
fun to_unsigned(n: Int): Int {
    if (n < 0) {
        return ((MAX + (n).toLong()).toInt())
    }
    return n
}

fun from_unsigned(n: Int): Int {
    if (n >= HALF) {
        return (((n).toLong() - MAX).toInt())
    }
    return n
}

fun bit_and(a: Int, b: Int): Int {
    var x: Int = (a).toInt()
    var y: Int = (b).toInt()
    var res: Int = (0).toInt()
    var bit: Int = (1).toInt()
    var i: Int = (0).toInt()
    while (i < 32) {
        if (((Math.floorMod(x, 2)) == 1) && ((Math.floorMod(y, 2)) == 1)) {
            res = res + bit
        }
        x = x / 2
        y = y / 2
        bit = bit * 2
        i = i + 1
    }
    return res
}

fun bit_xor(a: Int, b: Int): Int {
    var x: Int = (a).toInt()
    var y: Int = (b).toInt()
    var res: Int = (0).toInt()
    var bit: Int = (1).toInt()
    var i: Int = (0).toInt()
    while (i < 32) {
        var abit: Int = (Math.floorMod(x, 2)).toInt()
        var bbit: Int = (Math.floorMod(y, 2)).toInt()
        if ((Math.floorMod((abit + bbit), 2)) == 1) {
            res = res + bit
        }
        x = x / 2
        y = y / 2
        bit = bit * 2
        i = i + 1
    }
    return res
}

fun lshift1(num: Int): Int {
    return ((Math.floorMod(((num * 2).toLong()), MAX)).toInt())
}

fun add(a: Int, b: Int): Int {
    var first: Int = (to_unsigned(a)).toInt()
    var second: Int = (to_unsigned(b)).toInt()
    while (second != 0) {
        var carry: Int = (bit_and(first, second)).toInt()
        first = bit_xor(first, second)
        second = lshift1(carry)
    }
    var result: Int = (from_unsigned(first)).toInt()
    return result
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        println(_numToStr(add(3, 5)))
        println(_numToStr(add(13, 5)))
        println(_numToStr(add(0 - 7, 2)))
        println(_numToStr(add(0, 0 - 7)))
        println(_numToStr(add(0 - 321, 0)))
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}

File: tests/algorithms/x/PHP/maths/addition_without_arithmetic.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _intdiv($a, $b) {
    if ($b === 0 || $b === '0') {
        throw new DivisionByZeroError();
    }
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  $MAX = 4294967296;
  $HALF = 2147483648;
  function to_unsigned($n) {
  global $HALF, $MAX;
  if ($n < 0) {
  return $MAX + $n;
}
  return $n;
};
  function from_unsigned($n) {
  global $HALF, $MAX;
  if ($n >= $HALF) {
  return $n - $MAX;
}
  return $n;
};
  function bit_and($a, $b) {
  global $HALF, $MAX;
  $x = $a;
  $y = $b;
  $res = 0;
  $bit = 1;
  $i = 0;
  while ($i < 32) {
  if (($x % 2 == 1) && ($y % 2 == 1)) {
  $res = $res + $bit;
}
  $x = _intdiv($x, 2);
  $y = _intdiv($y, 2);
  $bit = $bit * 2;
  $i = $i + 1;
};
  return $res;
};
  function bit_xor($a, $b) {
  global $HALF, $MAX;
  $x = $a;
  $y = $b;
  $res = 0;
  $bit = 1;
  $i = 0;
  while ($i < 32) {
  $abit = $x % 2;
  $bbit = $y % 2;
  if (($abit + $bbit) % 2 == 1) {
  $res = $res + $bit;
}
  $x = _intdiv($x, 2);
  $y = _intdiv($y, 2);
  $bit = $bit * 2;
  $i = $i + 1;
};
  return $res;
};
  function lshift1($num) {
  global $HALF, $MAX;
  return ($num * 2) % $MAX;
};
  function add($a, $b) {
  global $HALF, $MAX;
  $first = to_unsigned($a);
  $second = to_unsigned($b);
  while ($second != 0) {
  $carry = bit_and($first, $second);
  $first = bit_xor($first, $second);
  $second = lshift1($carry);
};
  $result = from_unsigned($first);
  return $result;
};
  echo rtrim(_str(add(3, 5))), PHP_EOL;
  echo rtrim(_str(add(13, 5))), PHP_EOL;
  echo rtrim(_str(add(-7, 2))), PHP_EOL;
  echo rtrim(_str(add(0, -7))), PHP_EOL;
  echo rtrim(_str(add(-321, 0))), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
